### PR TITLE
Add local_files_only parameter to pretrained items

### LINKED ISSUE #2867: from_pretrained making internet connection if internet turned on
I'd like to ask why model.from_pretrained makes ssl connection event though I provide cache_dir? If I turn off the internet everything works just fine.

```
│     └─ 0.726 from_pretrained  transformers/tokenization_utils.py:256
│        └─ 0.726 _from_pretrained  transformers/tokenization_utils.py:311
│           ├─ 0.570 cached_path  transformers/file_utils.py:205
│           │  └─ 0.570 get_from_cache  transformers/file_utils.py:333
│           │     └─ 0.570 head  requests/api.py:91
│           │        └─ 0.570 request  requests/api.py:16
│           │           └─ 0.565 request  requests/sessions.py:466
│           │              └─ 0.558 send  requests/sessions.py:617
│           │                 └─ 0.558 send  requests/adapters.py:394
│           │                    └─ 0.557 urlopen  urllib3/connectionpool.py:494
│           │                       └─ 0.557 _make_request  urllib3/connectionpool.py:351
│           │                          ├─ 0.413 _validate_conn  urllib3/connectionpool.py:986
│           │                          │  └─ 0.413 connect  urllib3/connection.py:298
│           │                          │     ├─ 0.281 ssl_wrap_socket  urllib3/util/ssl_.py:296
│           │                          │     │  ├─ 0.263 wrap_socket  ssl.py:410
│           │                          │     │  │  └─ 0.263 _create  ssl.py:813
│           │                          │     │  │     └─ 0.263 do_handshake  ssl.py:1132
│           │                          │     │  └─ 0.018 [self]  
│           │                          │     └─ 0.132 _new_conn  urllib3/connection.py:143
│           │                          │        └─ 0.132 create_connection  urllib3/util/connection.py:33
│           │                          │           └─ 0.130 [self]  
│           │                          └─ 0.144 getresponse  http/client.py:1300
│           │                             └─ 0.144 begin  http/client.py:299
│           │                                └─ 0.144 _read_status  http/client.py:266
│           │                                   └─ 0.144 readinto  socket.py:575
│           │                                      └─ 0.144 recv_into  ssl.py:1060
│           │                                         └─ 0.144 read  ssl.py:920
```

and here's the output with internet turned off

```
     └─ 0.358 from_pretrained  transformers/tokenization_utils.py:256
│        └─ 0.358 _from_pretrained  transformers/tokenization_utils.py:311
│           ├─ 0.255 __init__  transformers/tokenization_bert.py:138
│           │  ├─ 0.163 load_vocab  transformers/tokenization_bert.py:98
│           │  │  └─ 0.160 [self]  
│           │  ├─ 0.056 <listcomp>  transformers/tokenization_bert.py:186
│           │  └─ 0.036 [self]  
│           └─ 0.102 cached_path  transformers/file_utils.py:205
│              └─ 0.101 get_from_cache  transformers/file_utils.py:333
│                 ├─ 0.083 head  requests/api.py:91
│                 │  └─ 0.083 request  requests/api.py:16
│                 │     └─ 0.080 request  requests/sessions.py:466
│                 │        ├─ 0.066 send  requests/sessions.py:617
│                 │        │  └─ 0.066 send  requests/adapters.py:394
│                 │        │     ├─ 0.046 urlopen  urllib3/connectionpool.py:494
│                 │        │     │  ├─ 0.035 _make_request  urllib3/connectionpool.py:351
│                 │        │     │  │  └─ 0.035 _validate_conn  urllib3/connectionpool.py:986
│                 │        │     │  │     └─ 0.035 connect  urllib3/connection.py:298
│                 │        │     │  │        └─ 0.035 _new_conn  urllib3/connection.py:143
│                 │        │     │  │           ├─ 0.015 create_connection  urllib3/util/connection.py:33
│                 │        │     │  │           │  └─ 0.014 getaddrinfo  socket.py:735
│                 │        │     │  │           ├─ 0.012 [self]  
│                 │        │     │  │           └─ 0.008 __init__  urllib3/exceptions.py:20
│                 │        │     │  └─ 0.006 increment  urllib3/util/retry.py:355
│                 │        │     ├─ 0.008 [self]  
│                 │        │     └─ 0.008 __init__  requests/exceptions.py:17
│                 │        ├─ 0.006 merge_environment_settings  requests/sessions.py:690
│                 │        │  └─ 0.005 get_environ_proxies  requests/utils.py:755
│                 │        └─ 0.006 [self]  
│                 └─ 0.014 filter  fnmatch.py:48
│                    └─ 0.009 _compile_pattern  fnmatch.py:38
│                       └─ 0.005 compile  re.py:232
│                          └─ 0.005 _compile  re.py:271
│                             └─ 0.005 compile  sre_compile.py:759
```

### PR DESCRIPTION
**Pull request by BramVanroy**
closes https://github.com/huggingface/transformers/issues/2867

Setting local_files_only=True disables outgoing traffic:
- etags are not looked up
- files are not downloaded (config, tokenizer, model)

An appropriate error is thrown when this argument may be the cause why a model cannot be loaded.

```python
import pyinstrument
from transformers import DistilBertConfig, DistilBertModel, DistilBertTokenizer


class TreeProfiler():
    def __init__(self, show_all=False):
        self.profiler = pyinstrument.Profiler()
        self.show_all = show_all # verbose output of pyinstrument profiler

    def __enter__(self):
        print("WITH TREE_PROFILER:")
        self.profiler.start()

    def __exit__(self, *args):
        self.profiler.stop()
        print(self.profiler.output_text(unicode=True, color=True, show_all=self.show_all))


def main():
    with TreeProfiler(show_all=True):
        config = DistilBertConfig.from_pretrained('distilbert-base-uncased', local_files_only=True)
        model = DistilBertModel.from_pretrained('distilbert-base-uncased', local_files_only=True)
        tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased', local_files_only=True)


if __name__ == '__main__':
    main()
```

The above snippet will throw an error message when the expected files are not present in the cache.  When they are, though, everything is loaded fine without the need of any additional lookups.

### CODE REVIEW

**Comment by julien-c** [comment]
I think this feature is reasonable but I'm not sure about the param name. Maybe something like `disable_networking` or `local_files_only`?

**Comment by BramVanroy** [comment]
Yeah, wasn't sure about a parameter name. I like `local_files_only`, even though it's quite long.

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -460,6 +463,7 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
             resolved_archive_file = None
 
         # Instantiate model.
+        print(model_kwargs)
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py]: Maybe remove this print?

Comment (BramVanroy) [reply src/transformers/modeling_utils.py]: @LysandreJik Good catch! Totally missed it.

**Review by LysandreJik — APPROVED** [review]
Cool, this is a nice addition and it's cleanly done.

**Comment by BramVanroy** [comment]
Would it make sense to utilise this in the examples? I am thinking about multi-GPU set-ups where the online lookup only has to be done by the main process (local_rank == 0). For all other processes, local_files_only can be True. Might avoid some redundant look-ups - even though in practice it won't matter much in terms of speed (couple of seconds at most).

**Review by sshleifer — APPROVED** [review]
LGTM, let's merge!

**Review by julien-c — APPROVED** [review]

**Pull request merged**